elixir/lib/supermarket/model/shopping_cart.ex

```diff
@@ -0,0 +1,99 @@
+defmodule Supermarket.Model.ShoppingCart do
+  require IEx
+  alias Supermarket.Model.Discount
+  alias Supermarket.Model.ProductQuantity
+  alias Supermarket.Model.Receipt
+  alias Supermarket.Model.SupermarketCatalog
+
+  defstruct [:items, :product_quantities]
+
+  def new, do: %__MODULE__{items: [], product_quantities: %{}}
+
+  def add_item(cart, product) do
+    add_item_quantity(cart, product, 1.0)
+  end
+
+  def add_item_quantity(cart, product, quantity) do
+    cart
+    |> Map.update!(:items, &[ProductQuantity.new(product, quantity) | &1])
+    |> Map.update!(:product_quantities, fn product_quantities ->
+      if Map.has_key?(product_quantities, product) do
+        Map.put(product_quantities, product, product_quantities[product] + quantity)
+      else
+        Map.put(product_quantities, product, quantity)
+      end
+    end)
+  end
+
+  def handle_offers(cart, receipt, offers, catalog) do
+    cart.product_quantities
+    |> Map.keys()
+    |> Enum.reduce(receipt, fn p, receipt ->
+      quantity = cart.product_quantities[p]
+
+      if Map.has_key?(offers, p) do
+        offer = offers[p]
+        unit_price = SupermarketCatalog.get_unit_price(catalog, p)
+        quantity_as_int = trunc(quantity)
+        discount = nil
+        x = 1
+
+        {discount, x} =
```

Comment:
Unfortunately I do not know much about Elexir.... Is the `cond` expression strictly necessary here? In the other languages this is the first step people have to consolidate the code which is more imperative and having different updates in the branches.

Reply:
The issue is that in Elixir variables are immutable (but can be rebound, and shadowed in nested scopes). So if we reproduced the shape of the java code along these lines (there’s no `elseif`, but `cond` is similar):

```elixir
x = 1
cond do
  offer.offer_type == :three_for_two) ->
    x = 3
  offer.offer_type == :two_for_amount ->
    x = 2
    # etc
end
```

… then `x` would always end up with the value of `1`, because the `x = 3` and `x = 2` are creating separate variables in their own block scopes. The only way to set a variable based on a conditional is to assign the result of the conditional to the variable.